### PR TITLE
Dpo penalty update

### PR DESCRIPTION
**Pull request by Eugene-hu**
- Adds additional checks for empty strings and smaller completions
- Adds a penalty for repeated tokens akin to huggingface's repeat penalty during generation (https://github.com/huggingface/transformers/blob/v4.32.0/src/transformers/generation/logits_process.py#L279) 
- The penalty will deter repeated tokens with subsequently heavier penalties each time it occurs.

### CODE REVIEW

File: openvalidators/reward/dpo.py [hunk 1]
```diff
@@ -33,17 +33,23 @@ def name(self) -> str: return RewardModelType.dpo.value
     def __init__(self, device: str):
         super().__init__()
         self.device = device
+        self.penalty = 1.2
```
Comment (shibshib) [inline openvalidators/reward/dpo.py:36]: do we have any data for why we picked 1.2 for penalty?

**Review by shibshib — APPROVED** [review]

File: openvalidators/reward/dpo.py (hunk 1 above)
Comment (Eugene-hu) [reply openvalidators/reward/dpo.py:36]: It is the same default parameter used by huggingface and was retrieved from this paper (https://arxiv.org/pdf/2305.14314.pdf)

File: openvalidators/reward/dpo.py [hunk 2]
```diff
@@ -33,17 +33,23 @@ def name(self) -> str: return RewardModelType.dpo.value
...
+    def reward_single(self, prompt: str, completion: str, name: str ,with_penalty=True) -> float:
         r""" Calculates a direct preference optimization (DPO) style reward for a completion,
         which is a reference model's average log-probability for completion tokens given a prompt.
         Uses guidance from https://github.com/eric-mitchell/direct-preference-optimization/blob/main/trainers.py.
         """
         with torch.no_grad():
+
+            # Check if completion is 
+            if completion.strip() == '' or len(completion) <= 5:
+                return -11 # exp(-11)=1.67e-5 < 2e-5=1/50257 (typical vocab size)
```
Comment (p-ferreira) [inline openvalidators/reward/dpo.py:51]: I'm not sure if I got it why is -11, could you please elaborate more so I could better understand it?

**Review by p-ferreira — APPROVED** [review]

File: openvalidators/reward/dpo.py (hunk 1 above)
Comment (p-ferreira) [reply openvalidators/reward/dpo.py:36]: I think that adding this reference to the code in one comment line could help clarify future doubts

File: openvalidators/reward/dpo.py (hunk 2 above)
Comment (Eugene-hu) [reply openvalidators/reward/dpo.py:51]: `exp(-11)` corresponds to base value given to zero or short responses; it is the nearest integer value that is less than equal probability value across all logits (1/50257).

Comment (p-ferreira) [reply openvalidators/reward/dpo.py:51]: Would it be feasible to calculate this in runtime by getting something like `1 / model.vocab_size`? That way the code will be independent of the model used as it would be calculated dynamically.

Comment (opentaco) [reply openvalidators/reward/dpo.py:51]: > Would it be feasible to calculate this in runtime by getting something like `1 / model.vocab_size`? That way the code will be independent of the model used as it would be calculated dynamically.

Yes this can be done in a future update, and will be necessary if the DPO model tokenizer is changed to something non-standard.

Comment (p-ferreira) [reply openvalidators/reward/dpo.py:51]: Got it, I will create an issue for that so we don't lose track of this
